setup - virtualization facts add Red Hat vendor

mnecas opened the pull request:
##### SUMMARY
<!--- Describe the change below, including rationale and design decisions -->
There were some changes in the RHV engine that changed sys_vendor to `Red Hat` this patch fixes the issue.
<!--- HINT: Include "Fixes #nnn" if you are fixing an existing issue -->
Fixes #72875
##### ISSUE TYPE
<!--- Pick one below and delete the rest -->
- Bugfix Pull Request

##### COMPONENT NAME
<!--- Write the short name of the module, plugin, task or feature below -->
setup facts
##### ADDITIONAL INFORMATION
<!--- Include additional information to help people understand the change here -->
<!--- A step-by-step reproduction of the problem is helpful if there is no related issue -->

<!--- Paste verbatim command output below, e.g. before and after your change -->
```paste below

```

Akasurde commented on lib/ansible/module_utils/facts/virtual/linux.py:
```diff
@@ -186,7 +186,7 @@ def get_virtual_facts(self):
         sys_vendor = get_file_content('/sys/devices/virtual/dmi/id/sys_vendor')
 
-        KVM_SYS_VENDORS = ('QEMU', 'oVirt', 'Amazon EC2', 'DigitalOcean', 'Google', 'Scaleway', 'Nutanix')
+        KVM_SYS_VENDORS = ('QEMU', 'oVirt', 'Red Hat', 'Amazon EC2', 'DigitalOcean', 'Google', 'Scaleway', 'Nutanix')
```
Could you please add a change log entry for this change? Thanks.

mwperina replied:
Please remove that line, for RHV detected we have code above, oVirt should be detected as kvm guest

mnecas commented:
Please do not merge, I will update also the `ansible_virtualization_type` to be RHV

mnecas commented:
- The RHV 4.4.3 chances product_name of the VM to RHEL
- RHEV is the old name for RHV (under 4.0) so switching to RHV, but also want to keep the RHEV for backward compatibility even if the name is not valid for 4.0 and higher. There are two options
  - keep it as in patch and have 2 names even if one is not correct
  - use only RHEV

mwperina commented on lib/ansible/module_utils/facts/virtual/linux.py:
```diff
@@ -152,6 +152,12 @@ def get_virtual_facts(self):
                 virtual_facts['virtualization_type'] = 'RHEV'
                 found_virt = True
 
+        if product_name == 'RHEL':
```
This is wrong, it will require further discussion within RHV team

michalskrivanek replied:
yes, because "RHEL" also means regular RHEL VMs (e.g. virt-manager created), not RHV

mwperina replied:
Let's remove this whole condition, above code should be enough

michalskrivanek commented on lib/ansible/module_utils/facts/virtual/linux.py:
```diff
@@ -152,6 +152,12 @@ def get_virtual_facts(self):
                 virtual_facts['virtualization_type'] = 'RHEV'
```
TBH I don't understand the original reasoning. "RHEV Hypervisor" is just a Red Hat's branded oVirt Node. It doesn't cover EL hosts at all that can be RHV too. It doesn't detect oVirt either, which is identified as kvm later by line 196 through Manufacturer field.

The fact should be called oVirt, as I do not see a downstream and upstream differentiation for any other product.
If we really need to distinguish them then it should use Manufacturer field. 

IMO we should be checking oVirt Manufacturer(here as bios_vendor) and "RHEV Hypervisor" product(for RHV<4.4), and System Family field (oVirt or RHV, to distinguish from RHEL and RHEL-AV plain virtualization). Earlier RHEL-based RHV hosts would still be misidentified as "kvm"  but that's the current behavior so it should be ok. 

See the logic since 4.4: https://gerrit.ovirt.org/#/c/ovirt-engine/+/106450/12/backend/manager/modules/vdsbroker/src/main/java/org/ovirt/engine/core/vdsbroker/builder/vminfo/LibvirtVmXmlBuilder.java

mnecas replied:
I want to distinguish them because of the backward compatibility otherways I would agree to have the same product name. Also not sure if I should keep the RHEV or use proper RHV.

Don't understand how we could use the bios_vendor to recognize the version.

As I understand it we could use product_name or the System Family to determine the type and approximate version of the engine.

(System Family - RHV<4.4 = Red Hat Enterprise Linux, RHV>4.4 = RHV)

michalskrivanek replied:
the question is if you need to keep bw compatibility even when it was wrong.
If you do, then just add a new condition ant type oVirt for 4.4+ cluster levels, and then I'd just do the check for System Family being RHV or oVirt.

You cannot use RHEL product name as that is the same for plain RHEL VMs. You would need further heuristics to distinguish old RHV on RHEL...and IMHO it's not worth getting into that mess

mwperina replied:
Let's modify the whole condition to:

```
    if product_name in ['RHV', 'RHEV Hypervisor']:
        guest_tech.add('RHV')
        if not found_virt:
            virtual_facts['virtualization_type'] = 'RHV'
            found_virt = True
```

mnecas replied:
I have added the product_family because the product_name is RHEL for the latest RHV.

mwperina reviewed (changes requested):
Please also perform below changes:

https://github.com/ansible/ansible/blob/devel/lib/ansible/module_utils/facts/virtual/linux.py#L313
```
                            if comm_content in ('vdsm', 'vdsmd'):
                                # We add both kvm and RHV to host_tech in this case.
                                # It's accurate. RHV uses KVM.
                                host_tech.add('RHV')
                                if not found_virt:
                                    virtual_facts['virtualization_type'] = 'RHV'
                                break
```

mwperina reviewed (approved):
+1

michalskrivanek commented:
@mwperina "oVirt should be detected as kvm guest" - why?
I mean, we did ignore backward compatibility and now report RHV for RHV (which is fine, I do believe it's best to fix this), but at the same time you keep oVirt reported as generic "kvm" - what's the point of that detection then? 
Somebody did want to identify RHEV at some point, are you sure the intent is/was to ignore oVirt? Even if most of the other checks are for upstream projects rather than individual products?

mwperina commented on lib/ansible/module_utils/facts/virtual/linux.py:
```diff
@@ -139,17 +139,24 @@ def get_virtual_facts(self):
 
-        if product_name == 'RHEV Hypervisor':
-            guest_tech.add('RHEV')
+        if product_name == 'oVirt Node':
```
Hmm, that's only for node hosts, but shouldn't we add below also to cover CentOS hosts?

```
if product_name == 'oVirt Node' or sys_vendor == 'oVirt':
...
```

mwperina reviewed (approved):
LGTM

mwperina reviewed (approved):
LGTM

michalskrivanek commented:
I would prefer to have only oVirt, but since there was a RHV before this is probably the best way...LGTM

Akasurde reviewed (approved):
LGTM

Akasurde reviewed (changes requested):
Still need to decide about 'RHEV'

mwperina commented:
> Still need to decide about 'RHEV'

RHEV has been renamed to RHV in RHV 4.0, which was released in August 2016, so the last version with RHEV name 3.6, which is completely unsupported from September 2018. So from my point of view we don't need to worry about change from RHEV to RHV anymore

mnecas commented:
I have added a new commit which could support both rhv and rhev for backward compatibility in the worst case I can reset the commit.

mnecas commented:
I have updated the condition once again and now it works in this way:
**Upstream**
`virtualization_type` will always be `oVirt`

**Downstream**
4.3 engine - `virtualization_type` will always be `RHEV`
4.4 engine - `virtualization_type` will always be for 4.4 RHV be `RHV` (does not care about cluster version)

mnecas committed:
Update of virtualization_type for oVirt and RHV

Upstream oVirt
virtualization_type will now always be oVirt

Downstream RHV engine
4.3 - virtualization_type will always be RHEV
4.4 - virtualization_type will always be RHV (does not care about cluster version)

Signed-off-by: mnecas <necas.marty@gmail.com>

mnecas committed:
check if RHEL is product_name of Red Hat vendor

mnecas commented:
> Shouldn't there be additional if:
> 
> ```
>  else if product_name == 'RHV':
> ```
> 
> Won't we wrongly get all Red Hat products as RHV without this if?

At that moment we would not have any `virtualization_type` for RHV 4.4 with cluster version 4.3. I have updated the check with `product_name == 'RHEL'`

mwperina commented:
> > Shouldn't there be additional if:
> > ```
> >  else if product_name == 'RHV':
> > ```
> > 
> > 
> > Won't we wrongly get all Red Hat products as RHV without this if?
> 
> At that moment we would not have any `virtualization_type` for RHV 4.4 with cluster version 4.3. I have updated the check with `product_name == 'RHEL'`

Arik/Michal are you OK with above solution? Or are you going to fix the issue with VMs in 4.3 cluster started by RHV Manager 4.4 which cannot be identified as RHV VMs?

@ahadas @michalskrivanek

mwperina commented:
Attaching current status checked by Martin:
![oVirt-RHV-virtualization-type](https://user-images.githubusercontent.com/6696369/108508914-d4877f80-72bc-11eb-81a2-435b8329ff55.png)

samdoran commented:
!waffling core_review

michalskrivanek commented:
> Attaching current status checked by Martin:
> ![oVirt-RHV-virtualization-type](https://user-images.githubusercontent.com/6696369/108508914-d4877f80-72bc-11eb-81a2-435b8329ff55.png)

this needs to include both rhvh/node and rhel /centos hosts, there is a difference in identification

michalskrivanek commented on lib/ansible/module_utils/facts/virtual/linux.py:
```diff
@@ -139,19 +139,32 @@ def get_virtual_facts(self):
+                if not found_virt:
+                    virtual_facts['virtualization_type'] = 'RHEV'
+                    found_virt = True
+            elif product_name == 'RHEL':
```
IIUC this is still wrong, it will return "RHV" on plain RHEL virtualization. As I commented earlier you need to look at Family (/sys/devices/virtual/dmi/id/product_family) for 4.4+

ahadas replied:
yeah, and then we can check the family first (and return `RHV` if it's RHV) and only then the product name - I see that also with latest versions, VDSM reports `RHEV Hypervisor` for RHVH

mnecas replied:
so something like this?
```python
if sys_vendor == 'Red Hat':
    if product_name == 'RHEV Hypervisor' and product_family == 'Red Hat Enterprise Linux' :
        guest_tech.add('RHEV')
        ...
    elif product_family == 'RHV' or (product_family == 'Red Hat Enterprise Linux' and product_name == 'RHEL'):
        guest_tech.add('RHV')
        ...
```

ahadas replied:
Or:

```python
if sys_vendor == 'Red Hat':
    if product_family == 'RHV':
        guest_tech.add('RHV')
        ...
    elif product_family == 'RHEV Hypervisor':
        guest_tech.add('RHEV')
        ...
    # and then maybe:
    else:
        ovirt_guest_agent_exists = ...
        if ovirt_guest_agent_exists:
            guest_tech.add('RHEV')
             ...
```
In the long run this would always return RHV and in the meantime we'll return `RHEV` also for 3.6<RHV<4.4.6 but that's not that bad I think if we expect clients to keep processing `RHEV` as an output - I doubt someone really cares about the brand. Maybe the version can be important but I don't see it's returned elsewhere

mnecas replied:
Not sure how it is returned in this version of ansible but in 2.9 where we will backport this logic so it can go to customers it is like this.
https://github.com/ansible/ansible/blob/5a9f4f7bb395f4738d682c87e659267cc3929d76/lib/ansible/module_utils/facts/virtual/linux.py#L92-L94
And users use the value not sure how/why but we have opened bz and ticket on it https://bugzilla.redhat.com/show_bug.cgi?id=1904085

How/where I can check the guest agent?

ahadas replied:
Ah, by 'brand' I meant RHEV vs. RHV. So as from 4.4.6, assuming that's where we'll change the sysinfo settings to always include family=RHV, we'll return `RHV` and for lower versions, we'll return `RHEV` for RHVH as in the snippet.

@nyoxi , can you please suggest how to check for ovirt-guest-agent existence in RHV?

michalskrivanek replied:
>  So as from 4.4.6

it's really just a bug in CL 4.5 until the fix in 4.4.5. We have family RHV everywhere since 4.4 and for 4.4 CL it works fine. Both RHEL and RHVH.

ahadas replied:
> it's really just a bug in CL 4.5 until the fix in 4.4.5. We have family RHV everywhere since 4.4 and for 4.4 CL it works fine. Both RHEL and RHVH.

I think you're mixing SKU and product-family. We're ok with CL 4.5 - the SKU is not there but the family is
The problem is different - it's with 4.3 clusters in RHV 4.4 which is still pretty common it seems

ahadas replied:
Posted https://gerrit.ovirt.org/#/c/ovirt-engine/+/113781/

nyoxi replied:
> @nyoxi , can you please suggest how to check for ovirt-guest-agent existence in RHV?

That depends on what exactly you want to test here. You can:

 - look if ovirt-guest-agent-common RPM  (or ovirt-guest-agent RPM on EL6 and older) is installed
 - look if there is virtio channel for the agent
 - look if ovirt-guest-agent.service is up and running

But I agree with what Michal said bellow -- it is better to avoid guest agent related check unless absolutely necessary.

ahadas replied:
> > @nyoxi , can you please suggest how to check for ovirt-guest-agent existence in RHV?
> 
> That depends on what exactly you want to test here. You can:
> 
> * look if ovirt-guest-agent-common RPM  (or ovirt-guest-agent RPM on EL6 and older) is installed
> * look if there is virtio channel for the agent

Yeah, that would be the best check it seems

> * look if ovirt-guest-agent.service is up and running
> 
> But I agree with what Michal said bellow -- it is better to avoid guest agent related check unless absolutely necessary.

So it depends on what we see as "absolutely necessary" - this PR is about identifying VMs as RHV VMS. Without checking ovirt-guest-agent we can't really tell whether VMs that run on plain RHEL are part of RHV or not (on 4.2/4.3 clusters and probably also on RHV < 4.4 with plain RHEL hosts) - if we're ok with not identifying them as RHV VMs, that's one things. But if we do, I guess we'll need to check that or something else in their metadata that is unique to RHV..

Edit: actually no, the metadata is not good for that, since we don't have it from within the guest..

michalskrivanek replied:
yes, virtio socket is trivial to check for and it's indeed a great idea for <=4.3 at least...it can be just that one thing to look at.

mwperina replied:
Nobody is checking sockets here, can't we just fix product family set by RHV Manager regardless of cluster version used?

ahadas commented:
> Attaching current status checked by Martin:
> ![oVirt-RHV-virtualization-type](https://user-images.githubusercontent.com/6696369/108508914-d4877f80-72bc-11eb-81a2-435b8329ff55.png)

Note that as from RHV 4.4.5, we should get product=RHEL-AV for 4.4 clusters

ahadas commented:
> > > Shouldn't there be additional if:
> > > ```
> > >  else if product_name == 'RHV':
> > > ```
> > > 
> > > 
> > > Won't we wrongly get all Red Hat products as RHV without this if?
> > 
> > 
> > At that moment we would not have any `virtualization_type` for RHV 4.4 with cluster version 4.3. I have updated the check with `product_name == 'RHEL'`
> 
> Arik/Michal are you OK with above solution? Or are you going to fix the issue with VMs in 4.3 cluster started by RHV Manager 4.4 which cannot be identified as RHV VMs?
> 
> @ahadas @michalskrivanek

I think we can always set the family to RHV on downstream, regardless of the cluster level and as Michal pointed out inline -  check the product family here instead of the product name. @michalskrivanek is there a reason this wouldn't work?

ahadas commented on lib/ansible/module_utils/facts/virtual/linux.py:
```diff
@@ -139,19 +139,32 @@ def get_virtual_facts(self):
+                guest_tech.add('RHV')
+                if not found_virt:
+                    virtual_facts['virtualization_type'] = 'RHV'
+                    found_virt = True
```
we can also consider checking if ovirt-guest-agent runs (as we check vdsmd on the host) to cover VMs that started on RHV 4.3 and keep running on 4.3 cluster in RHV 4.4

michalskrivanek replied:
I would avoid agent checks...if we can figuer out everything based on static immutable smbios data then let's do it that way. it's really not that complicated, the differences between versions are finite (sometimes it's also just a bug on ovirt side - like https://bugzilla.redhat.com/show_bug.cgi?id=1931786)

michalskrivanek replied:
though, right, the only case where smbios data is not good enough is 4.3 rhel host where it looks exactly the same as on plain virt (martin's table for rhv 4.4 cluster level 4.3)
I guess I would rather leave it be and identify as KVM and consider it a "bug" fixed in later CL

ahadas replied:
I don't mind much but if it's not too complicated then I would check for ovirt-guest-agent in a lazy way, meaning that we first check for product family (assuming we'll set it to all VMs that are started by RHVM 4.4) and if it's not set, check if ovirt-guest-agent is installed to cover everything. if it's complicated to do then yeah, I'd skip the last part

mwperina replied:
Can't do a change in RHV Manager 4.4, which would set proper product family to RHV regardess of cluster version? Meaning that even 4.2 VMs would have product family set correctly. As RHV Manager 4.2/4.3 are already out of support, I'd not care that 4.2 VM spawned by 4.2 RHVM has different product family than spawned by 4.4 RHVM

michalskrivanek replied:
it's not out of support, it's out of development, but it's still perfectly usable. I'd really want to avoid change of guest ABI in existing versions. That virtio socket check is trivial, just a file existence...

mnecas replied:
What file would we need to check?

michalskrivanek replied:
/dev/virtio-ports/ovirt-guest-agent.0 would be for OGA but we dropped it in el8/ubuntu18 in RHV 4.3 already I think.
/org.qemu.guest_agent.0 we add always...but @nyoxi are we the only ones using it with this name?
there's also /com.redhat.spice.0 but only when SPICE is defined, that is still a default.

ahadas replied:
Martin, to your question, there are really two approaches here:
1. the conservative approach that says VMs on 4.3 cluster-level with RHVM 4.4 should be exactly as if they were initiated by RHVM 4.3. this means those VMs should still be set with product-family = RHEL.
2. the progressive approach that considers exceptions to this rule when they make sense. for example, we removed the OGA channel also for RHEL 8 guests also on 4.3 cluster-level with RHVM 4.4. in theory it broke the ABI but no one really cares..

I tend to think we can go with the progressive approach here (because otherwise it would have raised complains by customers that upgraded 4.3 clusters to 4.4 cluster-level) - and then checking product-family covers all VMs that are started by RHVM 4.4 (regardless of their cluster level) and the check for OGA channel (which was dropped in 4.4.2 btw - https://bugzilla.redhat.com/show_bug.cgi?id=1845591) covers most of the VMs that were started by RHVM < 4.4.

but we didn't reach an agreement on the ability to take the progressive approach here. the only winning argument I can see for this is showing that changing the product family has no real impact (like dropping OGA channel on RHEL 8 guest in 4.3 cluster level) but unfortunately I can't prove that.
so I guess we can fall back to the conservative approach which leaves us with the identification problem for VMs on 4.3 cluster-level and keep exploring options like Michal suggested in his last comment.

nyoxi replied:
> /dev/virtio-ports/ovirt-guest-agent.0 would be for OGA but we dropped it in el8/ubuntu18 in RHV 4.3 already I think.

From what I have found it seems that the channel name was:

* before CL 4.2 -- named `com.redhat.rhevm.vdsm`
* in cluster level 4.2 -- renamed to `ovirt-guest-agent.0`
* oVirt 4.4.2 -- removed completely from EL8 and CoreOS VMs (for all cluster levels)

> /org.qemu.guest_agent.0 we add always...but @nyoxi are we the only ones using it with this name?
there's also /com.redhat.spice.0 but only when SPICE is defined, that is still a default.

The names for spice and qemu-ga channels are de-facto standards. It is not specific to oVirt.

michalskrivanek replied:
yeah, so with that i would conclude that we have no good way on 4.2/4.3 and leave it broken on RHEL (and it does work on 4.2/4.3 RHVH at least)
And finally merge this, it's worse that we keep it broken entirely on a current version for 3 months now:)

mnecas committed:
use product_family for RHEV

mkrizek commented:
!waffling needs_revision

mnecas committed:
add product_family and update fragment

michalskrivanek reviewed (approved):
this is a good improvement for oVirt/RHV 4.4+ even if it doesn't address issues on earlier versions or cluster levels.
If it's not good enough we can have a followup with more elaborated checks(e.g. channel definition present would work for non-recent OSes)

mwperina commented on changelogs/fragments/72876-setup-facts-add-redhat-vendor.yml:
```diff
@@ -0,0 +1,4 @@
+bugfixes:
+- "Setup virtualization_facts - add RHV and oVirt type (https://github.com/ansible/ansible/pull/72876)."
+- "This change will fully work from the 4.4.6 oVirt engine, for lower versions, there was a change
```
The patch was abandoned, so please remove the sentence. I'd just mention that this works only oVirt/RHV 4.4

michalskrivanek replied:
This change will fully work for VMs in clusters at cluster level 4.4 or newer

mwperina reviewed (changes requested):
Other than doc fragment it looks good to me

mnecas committed:
update fragment

mwperina reviewed (approved):
LGTM

mwperina commented:
Abhijeet, are you OK with the change?

@Akasurde

Akasurde commented on changelogs/fragments/72876-setup-facts-add-redhat-vendor.yml:
```diff
@@ -0,0 +1,3 @@
+bugfixes:
+- "Setup virtualization_facts - add RHV and oVirt type (https://github.com/ansible/ansible/pull/72876)."
```
```suggestion
- "Setup virtualization_facts - add RHV and oVirt type. This change will fully work for VMs in clusters at cluster level 4.4 or newer. (https://github.com/ansible/ansible/pull/72876)."
```

Akasurde replied:
Rest LGTM

Akasurde commented on changelogs/fragments/72876-setup-facts-add-redhat-vendor.yml:
```diff
@@ -0,0 +1,3 @@
+bugfixes:
+- "Setup virtualization_facts - add RHV and oVirt type (https://github.com/ansible/ansible/pull/72876)."
+- "This change will fully work for VMs in clusters at cluster level 4.4 or newer."
```
```suggestion
- "Setup virtualization_facts - add RHV and oVirt type. This change will fully work for VMs in clusters at cluster level 4.4 or newer (https://github.com/ansible/ansible/pull/72876)."

```

Akasurde replied:
Rest LGTM

mnecas committed:
Update changelogs/fragments/72876-setup-facts-add-redhat-vendor.yml

Co-authored-by: Abhijeet Kasurde <akasurde@redhat.com>

Akasurde reviewed (approved):
LGTM
